tuf/api/metadata.py

```diff
@@ -653,30 +652,61 @@ class VerificationResult:
     """Signature verification result for delegated role metadata.
 
     Attributes:
-        verified: True, if threshold of signatures is met.
-        signed: Set of delegated keyids, which validly signed.
-        unsigned: Set of delegated keyids, which did not validly sign.
+        threshold: Number of required signatures.
+        signed: dict of keyid:Key containing the keys that have signed.
+        unsigned: dict of keyid:Key containing the keys that have not signed.
+    """
+
+    threshold: int
+    signed: Dict[str, Key]
+    unsigned: Dict[str, Key]
+
+    def __bool__(self) -> bool:
+        return self.verified
+
+    @property
+    def verified(self) -> bool:
+        """True if threshold of signatures is met."""
+        return len(self.signed) >= self.threshold
+
+
+@dataclass
+class RootVerificationResult:
+    """Signature verification result for root metadata.
+
+    Root must be verified by itself and the previous root version. This
+    dataclass represents that combination. For the edge case of first version
+    of root, the same VerificationResult can be used twice.
+
+    Note that `signed` and `unsigned` correctness requires the underlying
+    VerificationResult keys to not conflict (no reusing the same keyid for
+    different keys).
 
+    Attributes:
+        first: First underlying VerificationResult
+        second: Second underlying VerificationResult
     """
 
-    verified: bool
-    signed: Set[str]
-    unsigned: Set[str]
+    first: VerificationResult
+    second: VerificationResult
 
     def __bool__(self) -> bool:
         return self.verified
 
-    def union(self, other: "VerificationResult") -> "VerificationResult":
-        """Combine two verification results.
+    @property
+    def verified(self) -> bool:
+        """True if threshold of signatures is met in both underlying VerificationResults."""
+        return self.first.verified and self.second.verified
 
-        Can be used to verify, if root metadata is signed by the threshold of
-        keys of previous root and the threshold of keys of itself.
-        """
-        return VerificationResult(
-            self.verified and other.verified,
-            self.signed | other.signed,
-            self.unsigned | other.unsigned,
-        )
+    @property
+    def signed(self) -> Dict[str, Key]:
+        """Dictionary of all signing keys that have signed, from both VerificationResults"""
+        return self.first.signed | self.second.signed
```

Comment:
this unfortunately requires python 3.9

Following likely does the same with dict unpacking but is even less readable
`return {**self.first.signed, **self.second.signed}`